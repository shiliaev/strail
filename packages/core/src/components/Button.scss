.sl-Button {
  display: inline-flex;
  padding: var(--spacing-m) var(--spacing-xl);
  align-items: center;
  justify-content: center;
  outline: none;
  border: none;
  cursor: pointer;
  border-radius: var(--border-radius-m);
  font-family: var(--font-family-roboto);
  font-size: var(--font-size-body-medium-desktop);
}

.sl-Button_size_small {
  height: var(--button-size-small);
  font-size: var(--font-size-body-small-desktop);
}

.sl-Button_size_medium {
  height: var(--button-size-medium);
  padding: var(--spacing-m) var(--spacing-2xl);
}

.sl-Button_size_large {
  height: var(--button-size-large);
  padding: var(--spacing-m) var(--spacing-3xl);
}

.sl-Button_variant_primary {
  background: var(--color-primary-400);
  color: #fff;
  transition: background ease-out .1s;

  &:hover,
  &:focus,
  &:active {
    background: var(--color-primary-300);
  }
}

.sl-Button_variant_secondary {
  background: transparent;
  color: var(--color-primary-300);
  border: 1px solid currentColor;
}

.sl-Button_disabled {
  cursor: not-allowed;
}
