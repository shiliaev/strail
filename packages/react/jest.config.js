module.exports = {
    testEnvironment: "jsdom",
    roots: ["<rootDir>/src"],
    transform: {
        "^.+\\.tsx?$": "ts-jest"
    },
    setupFilesAfterEnv: [
        // "@testing-library/react/cleanup-after-each", todo throws error
        "@testing-library/jest-dom/extend-expect"
    ],
    testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.tsx?$",
    moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
    moduleNameMapper: {
        "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.js",
        "\\.(css|scss)$": "identity-obj-proxy"
    }
};
