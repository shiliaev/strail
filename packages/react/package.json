{
  "name": "@strail/react",
  "sideEffects": false,
  "version": "0.0.1",
  "license": "MIT",
  "module": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook",
    "build:declarations": "tsc --jsx react-jsx --declaration --emitDeclarationOnly --outdir lib src/components/index.ts",
    "build:components": "yarn esbuild src/components/index.ts --bundle --outdir=lib --external:'react' --external:'clsx' --target=esnext --allow-overwrite --format=esm --jsx=transform",
    "build": "run-p build:*",
    "watch": "yarn esbuild src/components/index.ts --bundle --watch --outdir=lib --external:'react' --external:'clsx' --target=esnext --allow-overwrite --jsx-dev --format=esm",
    "lint": "eslint src/**/*.{ts,tsx} && prettier src --check",
    "lint:fix": "eslint src/**/*.{ts,tsx} --fix && prettier src --write"
  },
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib"
  ],
  "dependencies": {
    "@storybook/builder-vite": "0.2.4",
    "css-loader": "5.2.6",
    "esbuild": "0.14.51",
    "normalize.css": "8.0.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "sass": "1.55.0",
    "sass-loader": "10.1.1",
    "style-loader": "2.0.0",
    "typescript": "4.7.4",
    "vite": "3.0.4"
  },
  "devDependencies": {
    "@babel/core": "^7.16.5",
    "@storybook/addon-a11y": "6.5.12",
    "@storybook/addon-actions": "^6.5.12",
    "@storybook/addon-essentials": "^6.5.12",
    "@storybook/addon-links": "^6.5.12",
    "@storybook/addons": "6.5.12",
    "@storybook/builder-webpack5": "^6.5.12",
    "@storybook/manager-webpack5": "^6.5.12",
    "@storybook/preset-scss": "^1.0.3",
    "@storybook/react": "^6.5.12",
    "@storybook/theming": "6.5.12",
    "@testing-library/dom": "^8.10.1",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^12.1.2",
    "@types/jest": "^27.0.2",
    "@types/react": "^17.0.37",
    "@types/react-dom": "^17.0.10",
    "@types/react-test-renderer": "^17.0.1",
    "babel-loader": "^8.2.3",
    "clsx": "^1.1.1",
    "css-loader": "^6.5.1",
    "cypress": "^8.7.0",
    "identity-obj-proxy": "^3.0.0",
    "jest": "^27.3.1",
    "react-polymorphic-box": "^3.0.3",
    "react-test-renderer": "^17.0.2",
    "sass-loader": "^12.4.0",
    "style-loader": "^3.3.1",
    "ts-jest": "^27.0.7"
  },
  "peerDependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }
}
