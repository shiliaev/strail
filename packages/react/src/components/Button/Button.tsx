import clsx from "clsx";
import React, {ReactNode, ElementType, useRef} from "react";
import { Box, PolymorphicComponentProps } from "react-polymorphic-box";
import {Ripple} from "../Ripple/Ripple";

export enum ButtonSize {
  small = "small",
  medium = "medium",
  large = "large",
}

export enum ButtonVariant {
  primary = "primary",
  secondary = "secondary",
}

export interface ButtonOwnProps {
  className?: string;
  disabled?: boolean;
  children?: ReactNode;
  size?: ButtonSize;
  variant?: ButtonVariant;
  left?: ReactNode;
  right?: ReactNode;
}

const defaultElement = "button";
export type ButtonProps<E extends ElementType = typeof defaultElement> =
  PolymorphicComponentProps<E, ButtonOwnProps>;
export function Button<E extends ElementType = typeof defaultElement>({
  className,
  children,
  disabled,
  size = ButtonSize.medium,
  variant = ButtonVariant.primary,
  left,
  right,
  ...props
}: ButtonProps<E>) {
  const boxRef = useRef(null);

  return (
    <Box
      ref={boxRef}
      as={defaultElement}
      className={clsx(className, "sl-Button", {
        "sl-Button_disabled": disabled,
        [`sl-Button_variant_${variant}`]: variant,
        [`sl-Button_size_${size}`]: size,
      })}
      {...props}
    >
      <Ripple containerRef={boxRef} />
      {left && <span data-testid={"Button__left"}>{left}</span>}
      <span data-testid={"Button__children"}>{children}</span>
      {right && <span data-testid={"Button__right"}>{right}</span>}
    </Box>
  );
}

Button.Size = ButtonSize;
Button.Variant = ButtonVariant;
