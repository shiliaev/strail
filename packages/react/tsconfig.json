{
  "extends": "../../tsconfig.json",
  "include": [
    "./src/**/*.tsx",
    "./src/**/*.ts"
  ],
  "compilerOptions": {
    "outDir": "./lib",
    "declaration": true,
    "emitDeclarationOnly": true
  }
}
