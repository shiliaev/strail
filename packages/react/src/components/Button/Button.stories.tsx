import React from 'react';

import {Button, ButtonOwnProps, ButtonVariant} from './Button';

export default {
    component: Button,
};

//👇 We create a “template” of how args map to rendering
const Template = (args: ButtonOwnProps) => <Button {...args}>
    Button
</Button>;

export const Sandbox = Template.bind({});
