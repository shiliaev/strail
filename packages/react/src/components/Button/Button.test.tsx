import {render, fireEvent, screen, findByTestId} from '@testing-library/react'
import {Button} from "./Button";
import renderer from 'react-test-renderer'

const buttonText = 'Lorem ipsum dolor sit amet';
const customElementType = 'div';

describe('Button', () => {
    it('should has correct children text', async () => {
        render(<Button>{buttonText}</Button>);
        expect(screen.getByTestId('Button__children')).toHaveTextContent(buttonText);
    })

    it('should has correct element type', async () => {
        const {container}=render(<Button as={customElementType}>{buttonText}</Button>);
        expect(container.tagName).toMatch(customElementType.toUpperCase());
    })

    it('should call onClick when clicked', async () => {
        const onClick = jest.fn();
        render(<Button onClick={onClick}>{buttonText}</Button>);
        fireEvent.click(screen.getByText(buttonText));
        expect(onClick).toBeCalledTimes(1);
    })

    it('should has children', async () => {
        const tree = renderer.create(<Button><div/></Button>);
        expect(tree).toMatchSnapshot();
    })

    it('should has left slot', async () => {
        const tree = renderer.create(<Button left={<div />} />);
        expect(tree).toMatchSnapshot();
    })

    it('should has right slot', async () => {
        const tree = renderer.create(<Button right={<div />} />);
        expect(tree).toMatchSnapshot();
    })
})
