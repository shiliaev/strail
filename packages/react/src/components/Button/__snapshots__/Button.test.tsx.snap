// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has children 1`] = `
<div
  className="slButton"
>
  <span
    data-testid="Button__children"
  >
    <div />
  </span>
</div>
`;

exports[`has left slot 1`] = `
<div
  className="slButton"
>
  <span
    data-testid="Button__left"
  >
    <div />
  </span>
  <span
    data-testid="Button__children"
  />
</div>
`;

exports[`has right slot 1`] = `
<div
  className="slButton"
>
  <span
    data-testid="Button__children"
  />
  <span
    data-testid="Button__right"
  >
    <div />
  </span>
</div>
`;
